presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -381,12 +386,68 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
+        checkArgument(maxSplitsSizeInBytesPerPartition > 0,
+                "maxSplitsSizeInBytesPerPartition must be greater than zero: %s", maxSplitsSizeInBytesPerPartition);
+
+        boolean splitsDataSizeAvailable = splits.stream()
+                .allMatch(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().isPresent());
+        boolean autoTuneInitialPartitionCount = isSparkInitialPartitionCountAutoTuneEnabled(session) && splitsDataSizeAvailable;
+
+        int initialPartitionCount;
+        if (!autoTuneInitialPartitionCount) {
+            initialPartitionCount = getSparkInitialPartitionCount(session);
+            checkArgument(initialPartitionCount > 0, "initialPartitionCount must be greater then zero: %s", initialPartitionCount);
+        }
+        else {
+            long totalSizeInBytes = splits.stream()
+                    .mapToLong(split -> split.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong())
+                    .sum();
+            initialPartitionCount = max(1, toIntExact(totalSizeInBytes + maxSplitsSizeInBytesPerPartition - 1 / maxSplitsSizeInBytesPerPartition));
+        }
+
+        if (!splitsDataSizeAvailable) {
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                result.put(splitIndex % initialPartitionCount, splits.get(splitIndex));
+            }
+        }
+        else {
+            splits.sort((ScheduledSplit o1, ScheduledSplit o2) ->
+                    o1.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() >
+                            o2.getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong() ? -1 : 1);
+
+            PriorityQueue<SparkPartition> queue = new PriorityQueue();
+            int partitionCount = 0;
+
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                int partitionId;
+                long splitSizeInBytes = splits.get(splitIndex).getSplit().getConnectorSplit().getSplitSizeInBytes().getAsLong();
+
+                if (partitionCount >= initialPartitionCount &&
+                        !queue.isEmpty() &&
+                        queue.peek().getAvailableCapacityInBytes() >= splitSizeInBytes) {
+                    SparkPartition partition = queue.poll();
+                    partitionId = partition.getPartitionId();
+                    partition.assignSplitWithSize(splitSizeInBytes);
+                    if (partition.getAvailableCapacityInBytes() > 0) {
+                        queue.add(partition);
+                    }
+                }
+                else {
+                    partitionCount++;
+                    partitionId = partitionCount - 1;
+                    if (splitSizeInBytes < maxSplitsSizeInBytesPerPartition) {
+                        SparkPartition newPartition = new SparkPartition(partitionId, maxSplitsSizeInBytesPerPartition);
```

Comment:
Should this partition be placed into the queue?

Reply:
Not needed, this partition cannot accept more splits.

Reply:
Could you please elaborate how it works?

from what I see in the upper branch we check if `!queue.isEmpty()`, so if queue is empty it will not enter the upper branch. But we never place anything into the queue outside of the upper branch?

Reply:
With the redesign of original thought,  we removed `min_splits_per_partition`. The goal is now simple:   (1) Avoid skew by setting a max limit per partition. If a single split's data size exceeds this limit, the only thing we can do is to not add any more splits into this partition to make maximal partition size small.  (2) Make splits balanced, this is achieved by heuristic algorithm to always assign a split to partition with smallest size.

What is `partitionCount >= initialPartitionCount`?
-- `initialPartitionCount` is the lower bound of partitions (based on calculation total_size / max_partition_size), so if this is false, always create a new partition to accept any split

What is `!queue.isEmpty() && queue.peek().getAvailableCapacityInBytes() >= splitSizeInBytes`
-- If this condition is false,  there is no partition with enough capacity to accept this split, create a new partition.  

Invariance for partition in the queue
-- This partition is not full and contains at least one split